File: 백준/Silver/14501. 퇴사/퇴사.py
N = int(input())
arr = []

for i in range(N):
    T, P = map(int, input().split())
    arr.append((T, P))

dp = [0] * (N + 1)

for j in range(N-1, -1, -1):
    if j + arr[j][0] <= N:
        dp[j] = max(dp[j + arr[j][0]] + arr[j][1], dp[j + 1])
    else:
        dp[j] = dp[j + 1]

print(dp[0])

File: 백준/Silver/14501. 퇴사/퇴사.cc
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	int N;
	cin >> N;
	int dp[17] = { 0, };
	int arr[17][2] = { 0, };

	for (int i = 1;i <= N;i++) {
		cin >> arr[i][0] >> arr[i][1];
	}
	
	for (int i = N;i >= 1;i--) {
		if (i + arr[i][0] > N+1) {
			dp[i] = dp[i + 1];
		}
		else {
			dp[i] = max(dp[i + 1], dp[i + arr[i][0]] + arr[i][1]);
		}
	}

	cout << dp[1];

	return 0;
}